### PR TITLE
Feature/authentication

### PR DESCRIPTION
**Pull request by robinsdm**
Added the ability to create Applications. You can't modify them and they aren't authenticated yet.

### CODE REVIEW

File: app/io/pathfinder/pathfinderapplication/controllers/PathfinderApplicationController.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+                application.save();
+                return ok(Json.toJson(application));
+              } else {
+                return badRequest(response.asJson());
+              }
+            }
+          }
+      );
+
+      return resultPromise.get(30000);
```
Comment (aj-michael) [inline app/io/pathfinder/pathfinderapplication/controllers/PathfinderApplicationController.java]: magic number?

File: conf/application.conf [hunk 2]
```diff
@@ -11,7 +11,7 @@ play.crypto.secret = "changeme"
 # ~~~~~
 play.i18n.langs = [ "en" ]
 
-ebean.default=["io.pathfinder.authentication.models.*"]
+ebean.default=["io.pathfinder.authentication.models.*", "io.pathfinder.pathfinderapplication.models.*"]
```
Comment (aj-michael) [inline conf/application.conf]: What is the dfference between `io.pathfinder` and `io.pathfinder.pathfinderapplication`? In my opinion `io.pathfinder.pathfinderapplication` looks really ugly. Could we consolidate all models into `io.pathfinder.models` and all controllers into `io.pathfinder.controllers`?

**Comment by aj-michael** [comment]
Nice work david! :dog:

File: app/io/pathfinder/pathfinderapplication/controllers/PathfinderApplicationController.java (hunk 1 above)
Comment (robinsdm) [inline app/io/pathfinder/pathfinderapplication/controllers/PathfinderApplicationController.java]: Yeah, I just put that there to get it to work.

File: conf/application.conf (hunk 2 above)
Comment (robinsdm) [inline conf/application.conf]: I can change it. It was just to keep things separated.

**Comment by robinsdm** [comment]
It uses HTTP. There is no need for a socket request if it will only happen once and no one will need to be notified.

**Comment by robinsdm** [comment]
I'm going to get the conflict resolved and then resolve the problems above.

**Comment by aj-michael** [comment]
Ok maybe I'm misinterpreting class names, but isn't this line making a websocket connection?

```
WSRequest clusterCreateRequest = ws.url(pathfinderServerURL + "/cluster");
```

**Comment by robinsdm** [comment]
According to Play [WSRequest](https://www.playframework.com/documentation/2.4.x/JavaWS) is an HTTP request. Websockets are created with [WebSocket](https://www.playframework.com/documentation/2.4.x/JavaWebSockets). It probably stands for web service are something.

**Comment by robinsdm** [comment]
It should start building soon.

**Comment by robinsdm** [comment]
Also, I have no clue how to test the create application stuff. Since, it requires pathfinder-server to be running on port 9001.

**Comment by aj-michael** [comment]
`activator "~run 9001"` or hit v0.1.1 which is running on docker at 108.59.85.151:9000

**Comment by aj-michael** [comment]
Um David, you "resolved the conflict" by pushing all your code into dev, which bypasses this pull request.

**Comment by robinsdm** [comment]
Oh, sorry I meant no clue how to write tests for Travis to run. Git wouldn't let me see the conflicts on the feature branch. I got confused how to deal with it. All it needed was help with the conf/application.conf file.

**Comment by aj-michael** [comment]
There are two ways you should deal with something like that: merge dev into your branch (i dont like this method but some people do) or rebase your branch to start at the current tip of dev (i prefer this one, but if someone else has your branch checked out it messes them up)

**Comment by aj-michael** [comment]
Well in general, unit tests should never make network calls. The general practice is to use some kind of mocking or stubbing library that replaces whatever the thing is that you're using to make network requests.

**Comment by robinsdm** [comment]
Ok, we'll have to start looking into that. Play might have something.

**Comment by aj-michael** [comment]
@robinsdm: @danhanson and I were discussing application identifiers and we don't want them to be longs. Every web api that i've used has used some variant of UUIDs: https://en.wikipedia.org/wiki/Universally_unique_identifier.

I think we should generate these randomly when the user requests to create an application. I think this will make us look a little more professional.

**Comment by grovecj** [comment]
When I login, my password is returned to me in plain text as part of the user object. We should probably hide some of the data we send back.

**Comment by danhanson** [comment]
It is to help the user not forget it

**Pull request merged**